File: examples/go/case02_test.go
package examplesgo

import "testing"

func TestGenerationGateRejectsStaleAndAllowsCurrent(t *testing.T) {
	weak := &JobStore{}
	weak.Takeover(7)
	weak.Takeover(8) // B8 has not committed; state is still RUNNING.
	if !weak.StoreResultWeak("A7") || weak.Result != "A7" {
		t.Fatal("weak store did not admit the stale counterexample")
	}
	strong := &JobStore{}
	strong.Takeover(7)
	strong.Takeover(8)
	if strong.StoreResultGated(7, "A7") || strong.Result != "" {
		t.Fatal("stale generation changed protected result")
	}
	if !CommitCurrentWorker(strong, "B8") || strong.Result != "B8" {
		t.Fatal("current generation could not make progress")
	}
}

File: examples/python/case02_test.py
import unittest
from case02 import JobStore, commit_current_worker


class GenerationGateTest(unittest.TestCase):
    def test_stale_rejected_current_accepted(self):
        weak = JobStore()
        weak.takeover(8)  # B has not committed; still RUNNING.
        self.assertTrue(weak.store_result_weak("A7"))
        self.assertEqual(weak.result, "A7")
        gated = JobStore()
        gated.takeover(8)
        self.assertFalse(gated.store_result_gated(7, "A7"))
        self.assertIsNone(gated.result)
        self.assertTrue(commit_current_worker(gated, "B8"))
        self.assertEqual(gated.result, "B8")
